### PR TITLE
Index on vkey

### PR DESCRIPTION
**Pull request by swilly22**
Index entities on arrival
Allow for vkey capacity run-time config
Exit when encoding half-baked key on child process
Filter large number > 52 bits exact match index results

### CODE REVIEW

**Review by swilly22 — COMMENTED** [review]
@AviAvni Please see comments

File: tests/flow/test_index_scans.py [hunk 1]
```diff
@@ -527,3 +527,40 @@ def test18_index_scan_inside_apply(self):
...
+
+        result = redis_graph.query("MATCH (u:L1) WHERE u.id = 990000000262240069 OR u.id = 990000000262240070 RETURN u.id ORDER BY u.id")
+        expected_result = [[990000000262240069], [990000000262240070]]
+        self.env.assertEquals(result.result_set, expected_result)
+
+        result = redis_graph.query("MATCH (u1:L1), (u2:L1) WHERE u1.id = 990000000262240069 AND (u2.id = 990000000262240070 OR u2.id = 990000000262240071) RETURN u1.id, u2.id")
+        expected_result = [[990000000262240069, 990000000262240070], [990000000262240069, 990000000262240071]]
+        self.env.assertEquals(result.result_set, expected_result)
+
+        result = redis_graph.query("MATCH (u:L1) WITH min(u.id) as id MATCH (u1:L1), (u2:L1) WHERE u1.id = 990000000262240069 AND (u2.id = 990000000262240070 OR u2.id = 990000000262240071) RETURN u1.id, u2.id")
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply tests/flow/test_index_scans.py:556]: you wanted the 3 permutations I assumed they
1. IndexScanConsume
2. IndexScanConsumeFromChild + rebuild_index_query = false
3. IndexScanConsumeFromChild + rebuild_index_query = true

File: tests/flow/test_encode_decode.py [hunk 2]
```diff
@@ -177,3 +178,36 @@ def test08_multiple_reltypes(self):
...
+        self.env.assertEqual(response, "OK")
+
+        # Save RDB & Load from RDB
+        redis_con.save()
+        
+        log = logfile.read()
+
+        matches = re.findall("Created (.) virtual keys for graph vkey_max_entity_count", log)
+
+        self.env.assertEqual(matches, ['3', '6'])
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply tests/flow/test_encode_decode.py:213]: done

File: src/execution_plan/ops/op_index_scan.c [hunk 3]
```diff
@@ -191,20 +190,25 @@ static Record IndexScanConsume(OpBase *opBase) {
...
-	_UpdateRecord(op, r, *nodeId);
+	while((nodeId = RediSearch_ResultsIteratorNext(op->iter, op->idx, NULL))
+			!= NULL) {
+		// populate record with node
+		_UpdateRecord(op, r, *nodeId);
+		// apply unresolved filters
+		if(_PassUnresolvedFilters(op, r)) {
+			return r;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/execution_plan/ops/op_index_scan.c:210]: done

File: src/serializers/decoders/current/v10/decode_graph.c [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+	Graph_AllocateNodes(g, node_count);
+	Graph_AllocateEdges(g, edge_count);
+	for(uint64_t i = 0; i < label_count; i++) Graph_AddLabel(g);
+	for(uint64_t i = 0; i < relation_count; i++) Graph_AddRelationType(g);
+	// flush all matrices, guarantee matrix dimensions matches graph's nodes count
+	Graph_ApplyAllPending(g, true);
+}
+
+static GraphContext *_DecodeHeader(RedisModuleIO *rdb) {
+	/* Header format:
```
Comment (swilly22) [inline src/serializers/decoders/current/v10/decode_graph.c:42]: Please update this comment with the changes we've make to the format, make sure the encoder has a similar update to its comment.

File: src/serializers/decoders/current/v10/decode_graph_entities.c [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+		// * (labels) x M
+		// M will currently always be 0 or 1
+		uint64_t l = (nodeLabelCount) ? RedisModule_LoadUnsigned(rdb) : GRAPH_NO_LABEL;
+		Serializer_Graph_SetNode(gc->g, id, l, &n);
+
+		_RdbLoadEntity(rdb, gc, (GraphEntity *)&n);
+
+		if(l != GRAPH_NO_LABEL) {
+			Schema *s = GraphContext_GetSchemaByID(gc, l, SCHEMA_NODE);
```
Comment (swilly22) [inline src/serializers/decoders/current/v10/decode_graph_entities.c:105]: followed by `ASSERT(s != NULL);`

File: src/serializers/decoders/current/v10/decode_graph_schema.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	for(uint i = 0; i < index_count; i++) {
+		IndexType type = RedisModule_LoadUnsigned(rdb);
+		char *field = RedisModule_LoadStringBuffer(rdb, NULL);
+
+		if(!already_loaded)
+			Schema_AddIndex(&idx, s, field, type);
+		RedisModule_Free(field);
+	}
+	if(s) {
+		if(s->index) Index_Construct(s->index);
```
Comment (swilly22) [inline src/serializers/decoders/current/v10/decode_graph_schema.c:32]: Add comment: `no entities are expected to be in the graph in this point in time`
consider asserting this statement

File: src/serializers/decoders/current/v10/decode_graph.c [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+			case ENCODE_STATE_EDGES:
+				Graph_SetMatrixPolicy(gc->g, SYNC_POLICY_NOP);
+				RdbLoadEdges_v10(rdb, gc, payload.entities_count);
+				break;
+			case ENCODE_STATE_DELETED_EDGES:
+				RdbLoadDeletedEdges_v10(rdb, gc, payload.entities_count);
+				break;
+			case ENCODE_STATE_GRAPH_SCHEMA:
+				// skip handled in header
+				break;
```
[the remark this answers is not in the document]
Comment (AviAvni) [reply src/serializers/decoders/current/v10/decode_graph.c:168]: we can't remove it because we encoding all the states

**Pull request merged**